### PR TITLE
Use TensorFlow r2.2 in native client

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

**Comment by tu1258** [comment]
Hi, @lissyx 
Sorry to disturb you. Will remove my comment after you see it to keep comment section clean.
1. Will this be out soon(maybe 0.7.x)?
2. I see that you add `//tensorflow/core:rnn_ops_op_lib",  # BlockLSTM`. Could you add `//tensorflow/core:cudnn_rnn_ops_op_lib",  # cudnnLSTM` as well? Since it will allow user to train with cudnn and cudnnLSTM has more functions than BlockLSTM like dropout and num_proj.

Thanks!

**Comment by lissyx** [comment]
> Hi, @lissyx
> Sorry to disturb you. Will remove my comment after you see it to keep comment section clean.
> 
>     1. Will this be out soon(maybe 0.7.x)?
I hope it will be out soon, but it triggers weird behavior on CI under ARMv7 / NodeJS > 10 that are not yet understood, and I'm very busy with that and other things.

Will not be 0.7 for sure, this is a too big change for this.

> 
>     2. I see that you add `//tensorflow/core:rnn_ops_op_lib",  # BlockLSTM`. Could you add `//tensorflow/core:cudnn_rnn_ops_op_lib",  # cudnnLSTM` as well? Since it will allow user to train with cudnn and cudnnLSTM has more functions than BlockLSTM like dropout and num_proj.
> 

CuDNN training is already possible, and we already use BlockLSTM for a long time if you had a look at the code and the documentation.
Anyway, the code here is only to run inference, not for training.

> 
> Thanks!

**Comment by tu1258** [comment]
I get it. Thanks for your reply!

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Need devel version cause we need /usr/include/cudnn.h 
 # for compiling libctc_decoder_with_kenlm.so
-FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu18.04
+FROM nvidia/cuda:10.1-cudnn7-devel-ubuntu18.04
```
Comment (reuben) [inline Dockerfile:3]: Hm, is this going to be a problem for people using this Dockerfile to train? We probably need 10.0 and 10.1 side by side.

File: Dockerfile [hunk 2]
```diff
@@ -43,15 +43,12 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 
 RUN ln -s -f /usr/bin/python3 /usr/bin/python
 
-# Install NCCL 2.2
-RUN apt-get --no-install-recommends install -qq -y --allow-downgrades --allow-change-held-packages libnccl2=2.3.7-1+cuda10.0 libnccl-dev=2.3.7-1+cuda10.0
```
Comment (reuben) [inline Dockerfile:47]: Is this now part of the NVIDIA base image?

File: Dockerfile [hunk 3]
```diff
@@ -68,16 +65,18 @@ RUN wget https://bootstrap.pypa.io/get-pip.py && \
...
+ENV TF_NEED_OPENCL_SYCL 0
+ENV TF_NEED_OPENCL 0
 ENV TF_NEED_CUDA 1
-ENV TF_CUDA_PATHS "/usr/local/cuda,/usr/lib/x86_64-linux-gnu/"
-ENV TF_CUDA_VERSION 10.0
+ENV TF_CUDA_PATHS "/usr/include,/usr/local/cuda-10.1,/usr/lib/x86_64-linux-gnu/"
+ENV TF_CUDA_VERSION 10.1
 ENV TF_CUDNN_VERSION 7
 ENV TF_CUDA_COMPUTE_CAPABILITIES 6.0
-ENV TF_NCCL_VERSION 2.3
```
Comment (reuben) [inline Dockerfile:80]: Is this now picked up automatically?

File: Dockerfile [hunk 4]
```diff
@@ -125,11 +125,11 @@ RUN echo "build --spawn_strategy=standalone --genrule_strategy=standalone" \
...
-RUN mkdir /usr/local/cuda/lib &&  \
-    ln -s /usr/lib/x86_64-linux-gnu/libnccl.so.2 /usr/local/cuda/lib/libnccl.so.2 && \
-    ln -s /usr/include/nccl.h /usr/local/cuda/include/nccl.h && \
-    ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/local/cuda/lib64/stubs/libcuda.so.1 && \
-    ln -s /usr/include/cudnn.h /usr/local/cuda/include/cudnn.h
+##RUN mkdir /usr/local/cuda/lib &&  \
+##    ln -s /usr/lib/x86_64-linux-gnu/libnccl.so.2 /usr/local/cuda/lib/libnccl.so.2 && \
+##    ln -s /usr/include/nccl.h /usr/local/cuda/include/nccl.h && \
+##    ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/local/cuda/lib64/stubs/libcuda.so.1 && \
+##    ln -s /usr/include/cudnn.h /usr/local/cuda/include/cudnn.h
```
Comment (reuben) [inline Dockerfile:132]: Was leaving this commented out instead of removing it done on purpose?

File: doc/USING.rst [hunk 5]
```diff
@@ -28,7 +28,7 @@ Please refer to your system's documentation on how to install these dependencies
 CUDA dependency
 ^^^^^^^^^^^^^^^
 
-The GPU capable builds (Python, NodeJS, C++, etc) depend on the same CUDA runtime as upstream TensorFlow. Currently with TensorFlow 1.15 it depends on CUDA 10.0 and CuDNN v7.6. `See the TensorFlow documentation <https://www.tensorflow.org/install/gpu>`_.
+The GPU capable builds (Python, NodeJS, C++, etc) depend on the same CUDA runtime as upstream TensorFlow. Currently with TensorFlow r2.2 it depends on CUDA 10.1 and CuDNN v7.6. `See the TensorFlow documentation <https://www.tensorflow.org/install/gpu>`_.
```
Comment (reuben) [inline doc/USING.rst:31]: r2.2 -> 2.2

File: native_client/javascript/Makefile [hunk 6]
```diff
@@ -41,10 +41,6 @@ package.json: package.json.in
 		package.json.in > package.json && cat package.json
 
 npm-dev: package.json
-ifeq ($(findstring _NT,$(OS)),_NT)
-	# node-gyp@5.x behaves erratically with VS2015 and MSBuild.exe detection
-	$(NPM_TOOL) install node-gyp@4.x
-endif
```
Comment (reuben) [inline native_client/javascript/Makefile:47]: Nice.

Comment (reuben) [reply native_client/javascript/Makefile:47]: Should the node-gyp entry in devDependencies now be changed to just "5.x"?

File: taskcluster/pyenv-build.sh [hunk 7]
```diff
@@ -18,11 +18,7 @@ for pyver_conf in ${SUPPORTED_PYTHON_VERSIONS}; do
 
   pyalias="${pyver}_${pyconf}"
 
-  maybe_ssl102_py37 ${pyver}
-
-  LD_LIBRARY_PATH=${PY37_LDPATH}:$LD_LIBRARY_PATH \
-      PYTHON_CONFIGURE_OPTS="--enable-unicode=${pyconf} ${PY37_OPENSSL}" \
-      pyenv_install ${pyver} ${pyalias}
```
Comment (reuben) [inline taskcluster/pyenv-build.sh:25]: Nice

File: Dockerfile (hunk 1 above)
Comment (lissyx) [reply Dockerfile:3]: Why would it be a problem ?

File: Dockerfile (hunk 2 above)
Comment (lissyx) [reply Dockerfile:47]: It was there, just different version, as much as I remember.

File: Dockerfile (hunk 3 above)
Comment (lissyx) [reply Dockerfile:80]: Looks like it is

File: Dockerfile (hunk 1 above)
Comment (reuben) [reply Dockerfile:3]: Doesn't training with 1.15 require the older version and deps, like NCCL?

File: Dockerfile (hunk 4 above)
Comment (lissyx) [reply Dockerfile:132]: Yes, until we could confirm it was 100% okay

File: native_client/javascript/Makefile (hunk 6 above)
Comment (lissyx) [reply native_client/javascript/Makefile:47]: I think I did it, yes

File: Dockerfile (hunk 1 above)
Comment (lissyx) [reply Dockerfile:3]: That's not completely false, I guess we can't keep one dockerfile for everything?

Comment (reuben) [reply Dockerfile:3]: Yeah, would probably be better to have a simpler Dockerfile with just training stuff without building the world from scratch.
